src/NHibernate/Driver/DriverBase.cs

```diff
@@ -178,13 +178,23 @@ protected virtual SqlStringFormatter GetSqlStringFormatter()
 			return new SqlStringFormatter(this, ";");
 		}
 
-		private void SetCommandParameters(DbCommand cmd, SqlType[] sqlTypes)
+		private void SetCommandParameters(DbCommand cmd, SqlType[] sqlTypes, Parameter[] parameters)
 		{
-			for (int i = 0; i < sqlTypes.Length; i++)
+			var usedParameters = new HashSet<int>();
+			for (var i = 0; i < sqlTypes.Length; i++)
 			{
-				string paramName = ToParameterName(i);
+				var parameter = parameters.Length > i ? parameters[i] : null;
+				var position = parameter?.ParameterPosition ?? i;
```

Comment:
I am suspicious about this change but I have not yet had the time to analyze and understand its rationals. It is not obvious why `sqlTypes` ordering is supposed to match `parameters` one. (Maybe a comment at line 136 could explain it.)

I do not have yet seen either in which case two parameters may declare the same position (could it be three by the way?), in which case parameter positions may be different of `sqlTypes` indexes, ...

Reply:
From what I've seen by debugging, `parameters` were always ordered the same way as `sqlTypes` though there were some exceptions. For example when running [this](https://github.com/nhibernate/nhibernate-core/blob/1919480bff54c563902bdefabf462c26e4e45611/src/NHibernate.Test/NHSpecificTest/NH3634/FixtureByCode.cs#L92) test the number of `parameters` is two and `sqlTypes` is three, as the last parameter is `null`. There is [another special case](https://github.com/nhibernate/nhibernate-core/blob/ad4c2ef101cbf8ba798220973f4f78dd795e0896/src/NHibernate.Test/NHSpecificTest/NH1693/Fixture.cs#L6) that have duplicated parameters, where `ParameterPosition` is the same.
As I recently modified the batcher in order that only one command is created, this change is not needed anymore, so I will revert it in order to prevent having an unexpected breaking change.